### PR TITLE
Adding Loading Screen to User Sign-In

### LINKED ISSUE #790: Add Loading Screen On User Sign In
**Describe the solution you'd like**
When the user signs in, overlay a loading screen over the login page until signing in is complete (or while user is in "pending" state). For example, a loading screen can be a semi-transparent overlay over everything with a loading spinner in the center. 

**Additional context**
Currently, there is this code segment that causes the login page to be blank while the user is pending:
```
if (user !== undefined) {
    return <></>;
  }
```
This code logic should be replaced with the login loading screen code.

There may still be a blank page that appears after the loading screen appears once this is implemented btw. This could be because the login page automatically redirects to the dashboard page on sign in (which is also programmed to be blank while the user is pending). I was thinking of implementing skeleton loading for login-protected pages (dashboard, train, settings, etc) which would resolve that issue, but that likely will happen in a future pr once those pages have a clear layout structure. In short, if the blank page is still there for a brief moment but the loading screen shows up beforehand, there's no need to worry about that behavior for this pr.

**Setup Instructions (what branch to work off of)**
Run the following commands
```
git checkout nextjs
git pull origin nextjs
git checkout -b sign-in-loading-screen
```
FYI: If you are not able to immediately run `git checkout nextjs`, make sure you commit your changes in the current branch or run `git stash` and then execute the above commands

### PR DESCRIPTION
**Pull request by codingwithsurya**
# Adding Loading Screen to User Sign-In

What user problem are we solving?
Currently, when the user signs in, there is no visual indication that the authentication process is ongoing, leading to confusion and uncertainty for the users. This Pull Request aims to address this issue by overlaying a loading screen over the login page while the user is signing in or in the "pending" state. This loading screen will provide feedback to the users, assuring them that the sign-in process is underway.

What solution does this PR provide?
This PR introduces a loading screen that appears when the user clicks the sign-in button. The loading screen will be displayed until the authentication process is complete or until the user's state is no longer in the "pending" state. The loading screen will be a semi-transparent overlay over the entire login page, with a centered loading spinner.

Testing Methodology:

Tested on localhost

Any other considerations
As mentioned earlier, while this PR introduces the loading screen for the login page, there might be a momentary blank page issue when the user is redirected to the dashboard page after authentication. This behavior will be addressed in subsequent PRs by implementing skeleton loading for login-protected pages to provide a better user experience, as @dwu359 mentioned in the corresponding linked issue.

In summary, this PR lays the groundwork for the loading screen on the login page during the authentication process. The minor issue with the blank page during redirection will be tackled in future commits.

### CODE REVIEW

**Comment by karkir0003** [comment]
screen recording of feature @codingwithsurya

**Review by karkir0003 — COMMENTED** [review]
Have @farisdurrani or @dwu359 review in more detail. Please include screen recording of login action so we can see the overlay? @codingwithsurya

**Comment by codingwithsurya** [comment]
Here is the recording. As u can see, the loading screen is there on the click, but after user signs in there's a blank page, then briefly a login page, and then finally the dashboard. 
https://github.com/DSGT-DLP/Deep-Learning-Playground/assets/73920354/2afb1189-882f-4953-9f6f-496c908e4a28

**Comment by karkir0003** [comment]
is the "fix blank page showing momentarily after login" written in a separate github issue? @codingwithsurya

**Comment by codingwithsurya** [comment]
> is the "fix blank page showing momentarily after login" written in a separate github issue? @codingwithsurya

i dont think so. The login page automatically redirects to the dashboard page on sign in, which is programmed to be blank while the user is pending. 

Check out Daniel's thoughts on how we could resolve that issue in the linked issue description.

**Comment by karkir0003** [comment]
> > is the "fix blank page showing momentarily after login" written in a separate github issue? @codingwithsurya
> 
> i dont think so. The login page automatically redirects to the dashboard page on sign in, which is programmed to be blank while the user is pending. 
> 
> Check out Daniel's thoughts on how we could resolve that issue in the linked issue description. 

understood. thx

**Comment by karkir0003** [comment]
if faris or daniel approve, you can merge

**Review by farisdurrani — APPROVED** [review]

**Comment by farisdurrani** [comment]
Approved, @codingwithsurya go ahead and merge it in. We can make that skeleton loading page later but it isn't a concern since no protected content is being displayed

**Comment by codingwithsurya** [comment]
> Approved, @codingwithsurya go ahead and merge it in. We can make that skeleton loading page later but it isn't a concern since no protected content is being displayed

Thanks!
